### PR TITLE
Add session storage clustering support 

### LINKED ISSUE #3553: implement clustered memory store
supporting memcached and redis

github.com/eko/gocache

### PR DESCRIPTION
**Pull request by woutslakhorst**
closes #3553

- adds Memcached and Redis as options for session storage (Redis was already added but broken)
- added e2e tests for memcached and Redis sentinel

### CODE REVIEW

File: e2e-tests/clustering/memcached/node-A/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,71 @@
+load_module /usr/lib/nginx/modules/ngx_http_js_module.so;
```
Comment (reinkrul) [inline e2e-tests/clustering/memcached/node-A/nginx.conf:1]: we have "shared" HTTP config for e2e tests, can't we use that?

File: e2e-tests/clustering/redis/node-A/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,71 @@
+load_module /usr/lib/nginx/modules/ngx_http_js_module.so;
```
Comment (reinkrul) [inline e2e-tests/clustering/redis/node-A/nginx.conf:1]: these are all the same for all clustering tests? (maybe deduplicate them)

File: e2e-tests/clustering/redis/node-A/presentationexchangemapping.json [hunk 3]
```diff
@@ -0,0 +1,101 @@
+{
+  "test": {
+    "organization": {
+      "format": {
+        "ldp_vp": {
```
Comment (reinkrul) [inline e2e-tests/clustering/redis/node-A/presentationexchangemapping.json:5]: same for this (deduplicate)?

File: e2e-tests/clustering/redis/run-test.sh [hunk 4]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
+source ../../util.sh
+
+echo "------------------------------------"
```
Comment (reinkrul) [inline e2e-tests/clustering/redis/run-test.sh:4]: These are copies as well?

File: go.mod [hunk 5]
```diff
@@ -202,8 +202,19 @@ require (
 )
 
 require (
+	github.com/benbjohnson/clock v1.3.0 // indirect
+	github.com/bradfitz/gomemcache v0.0.0-20230905024940-24af94b03874 // indirect
+	github.com/daangn/minimemcached v1.2.0 // indirect
```
Comment (reinkrul) [inline go.mod:207]: this sounds like embedded memcached?

File: storage/engine.go [hunk 6]
```diff
@@ -188,6 +192,14 @@ func (e *engine) Configure(config core.ServerConfig) error {
...
 		}
 		e.sessionDatabase = NewRedisSessionDatabase(client, redisConfig.Database)
+		log.Logger().Info("Redis session storage support enabled.")
+	} else if memcachedConfig.isConfigured() {
+		memcachedClient, err := newMemcachedClient(memcachedConfig)
+		if err != nil {
+			return fmt.Errorf("unable to initialize memcached client: %w", err)
+		}
+		e.sessionDatabase = NewMemcachedSessionDatabase(memcachedClient)
+		log.Logger().Info("Memcached session storage support enabled.")
```
Comment (reinkrul) [inline storage/engine.go:202]: add this line for Redis as well (or remove it here) (and consider it for in0-memory).

File: storage/session.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+// StringOrBytes is a type that can be either a string or a byte slice
+// used for generic type constraints
+type StringOrBytes interface {
+	~string | ~[]byte
+}
+
+// SessionStoreImpl is an implementation of the SessionStore interface
+// It handles logic for all session store types
+type SessionStoreImpl[T StringOrBytes] struct {
+	underlying *cache.Cache[T]
```
Comment (reinkrul) [inline storage/session.go:47]: does everything go through the memcache API now? (in-memory as well?)

If not, this should be named something like MemcachedSessionStore?

**Comment by reinkrul** [comment]
Do we need/have something in diagnostics on the session store?

File: README.rst [hunk 8]
```diff
@@ -212,6 +212,7 @@ The following options can be configured on the server:
...
     **Storage**
+    storage.session.memcached.address             []                                                                                                                                                                                                                                                                                                                                                                                                                                                                       List of Memcached server addresses. These can be a simple 'host:port' or a Memcached connection URL with scheme, auth and other options.
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply README.rst:215]: or the entire connection string as `NUTS_` env variable?

File: docs/pages/deployment/storage.rst [hunk 9]
```diff
@@ -44,6 +44,68 @@ Refer to the documentation of the driver for the database you are using for the
     Usage of SQLite is not recommended for production environments.
     Connections to a SQLite DB are restricted to 1, which will lead to severe performance reduction.
 
+Session storage
+***************
+
+Session storage is used for storing access tokens, nonces and other volatile data.
+Data is stored in-memory only. There are 3 supported session storage types:
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply docs/pages/deployment/storage.rst:51]: should be `volatile`

File: go.mod (hunk 5 above)
Comment (woutslakhorst) [reply go.mod:207]: for test

File: storage/session.go (hunk 7 above)
Comment (woutslakhorst) [reply storage/session.go:47]: This is a generic impl, memcached/redis and in-memory initialize this with a different underlying cache. The lib used is a facade for the different caches.

File: storage/engine.go (hunk 6 above)
Comment (woutslakhorst) [reply storage/engine.go:202]: already is

**Comment by woutslakhorst** [comment]
> Do we need/have something in diagnostics on the session store?

there's no `ping` but maybe we can do a simple Get on a random value which should return ErrNotFound.

File: README.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (reinkrul) [reply README.rst:215]: Yeah, but we can still do it when someone actually requests it

File: storage/engine.go [hunk 10]
```diff
@@ -99,6 +99,18 @@ func (e *engine) CheckHealth() map[string]core.Health {
 		}
 		results["sql"] = sqlHealth
 	}
+	if e.sessionDatabase != nil {
+		results["session"] = core.Health{
+			Status: core.HealthStatusUp,
+		}
+		err := e.sessionDatabase.GetStore(defaultSessionDataTTL, "healthcheck").Get("does_not_exist", nil)
```
Comment (reinkrul) [inline storage/engine.go:106]: I hope so :-p

File: storage/session.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+var _ SessionStore = (*SessionStoreImpl[[]byte])(nil)
+
+// defaultSessionDataTTL is the default time to live for session data
+// some stores require a default
+var defaultSessionDataTTL = 15 * time.Minute
+
+// StringOrBytes is a type that can be either a string or a byte slice
+// used for generic type constraints
+type StringOrBytes interface {
+	~string | ~[]byte
```
Comment (reinkrul) [inline storage/session.go:41]: wow... is this Go?

File: storage/session_memcached.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+)
+
+type MemcachedSessionDatabase struct {
+	client     *memcache.Client
+	underlying *cache.Cache[[]byte]
+}
+
+// NewMemcachedSessionDatabase creates a new MemcachedSessionDatabase using an initialized memcache.Client.
+func NewMemcachedSessionDatabase(client *memcache.Client) *MemcachedSessionDatabase {
+	memcachedStore := memcachestore.NewMemcache(client, store.WithExpiration(defaultSessionDataTTL))
```
Comment (reinkrul) [inline storage/session_memcached.go:38]: Shouldn't we use the same value for access token lifetime? Or at least, the token lifetime shouldn't be longer than this default. Having 2 seeminly unrelated constants (but secretly relating) seems rather dangerous?

**Review by reinkrul — CHANGES_REQUESTED** [review]

File: storage/session_memcached.go (hunk 12 above)
Comment (woutslakhorst) [reply storage/session_memcached.go:40]: I added a comment. Memcached requires a default setting. We override this value when writing with the TTL of the store.

**Review by reinkrul — APPROVED** [review]

**Pull request merged**